### PR TITLE
Add `PartialDerivativeCoefficients` representation to lindiffops

### PR DESCRIPTION
**Pull request by timweiland**
Each linear differential operator can be represented as a sum of partial derivatives, and this PR adds this representation to every lindiffop.

EDIT: Rework done

### CODE REVIEW

File: src/linpde_gp/linfuncops/diffops/_coefficients.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    Parameters
+    ----------
+    coefficient_dict :
+        Dictionary of coefficients.
+    """
+
+    def __init__(
+        self, coefficient_dict: Dict[ShapeType, Dict[Tuple[int, ...], float]]
+    ) -> None:
+        self._input_codomain_shape_bound = None
```
Comment (marvinpfoertner) [inline src/linpde_gp/linfuncops/diffops/_coefficients.py:41]: I think we should pass `domain_shape` and `input_codomain_shape` as constructor parameters to ensure the integrity of the datastructure. That way we can also make `validate_input_{co}domain_shape` part of the constructor.

**Review by marvinpfoertner — COMMENTED** [review]
Nice work! 🙏🏻  Just some minor comments

File: src/linpde_gp/linfuncops/diffops/_directional_derivative.py [hunk 2]
```diff
@@ -6,14 +6,39 @@
...
 
 
 class DirectionalDerivative(LinearDifferentialOperator):
     def __init__(self, direction: ArrayLike):
-        self._direction = np.asarray(direction)
+        direction = np.asarray(direction)
+        if direction.ndim > 1:
+            raise ValueError("Direction must be element of R^n.")
+
+        def get_one_hot(index: int) -> np.ndarray:
```
Comment (marvinpfoertner) [inline src/linpde_gp/linfuncops/diffops/_directional_derivative.py:20]:
```suggestion
        def get_one_hot(index: int) -> tuple[int, ...]:
```

File: src/linpde_gp/linfuncops/diffops/_laplacian.py [hunk 3]
```diff
@@ -35,7 +35,24 @@ def __init__(self, weights: ArrayLike) -> None:
                 "most 1."
             )
 
-        super().__init__(input_shapes=(weights.shape, ()))
+        def get_one_hot(index: int) -> np.ndarray:
```
Comment (marvinpfoertner) [inline src/linpde_gp/linfuncops/diffops/_laplacian.py:38]:
```suggestion
        def get_one_hot(index: int) -> tuple[int, ...]:
```

File: src/linpde_gp/linfuncops/diffops/_coefficients.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
+from collections.abc import Mapping
+from copy import deepcopy
+from typing import Dict, Iterator, Tuple
```
Comment (marvinpfoertner) [inline src/linpde_gp/linfuncops/diffops/_coefficients.py:3]:
```suggestion
from collections.abc import Mapping, Iterator
from copy import deepcopy
```

File: src/linpde_gp/linfuncops/diffops/_coefficients.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+    Parameters
+    ----------
+    coefficient_dict :
+        Dictionary of coefficients.
+    """
+
+    def __init__(
+        self,
+        coefficient_dict: Dict[ShapeType, Dict[Tuple[int, ...], float]],
```
Comment (marvinpfoertner) [inline src/linpde_gp/linfuncops/diffops/_coefficients.py:43]:
```suggestion
        coefficient_dict: dict[ShapeType, dict[tuple[int, ...], float]],
```

File: src/linpde_gp/linfuncops/diffops/_coefficients.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+    @property
+    def input_domain_shape(self) -> ShapeType:
+        return self._input_domain_shape
+
+    @property
+    def input_codomain_shape(self) -> ShapeType:
+        return self._input_codomain_shape
+
+    def __getitem__(self, key: ShapeType) -> Dict[Tuple[ShapeType, int], float]:
```
Comment (marvinpfoertner) [inline src/linpde_gp/linfuncops/diffops/_coefficients.py:89]:
```suggestion
    def __getitem__(self, key: ShapeType) -> dict[tuple[int, ...], float]:
```

**Comment by timweiland** [comment]
Added a new `MultiIndex` class as discussed offline.

File: src/linpde_gp/linfuncops/diffops/_coefficients.py [hunk 7]
```diff
@@ -1,14 +1,61 @@
...
+    @classmethod
+    def from_one_hot(
+        cls, index: tuple[int, ...], shape: ShapeType, order: int
+    ) -> "MultiIndex":
+        multi_index = np.zeros(shape, dtype=int)
+        multi_index[index] = order
+        return cls(multi_index)
+
+    @functools.cached_property
+    def total_order(self) -> int:
```
Comment (marvinpfoertner) [inline src/linpde_gp/linfuncops/diffops/_coefficients.py:35]:
```suggestion
    def order(self) -> int:
```

File: src/linpde_gp/linfuncops/diffops/_coefficients.py [hunk 8]
```diff
@@ -1,14 +1,61 @@
...
+    """
+
+    def __init__(self, multi_index: ArrayLike) -> None:
+        self._multi_index = np.asarray(multi_index, dtype=int)
+        if np.any(self._multi_index < 0):
+            raise ValueError(f"Multi-index {multi_index} contains negative entries.")
+        self._multi_index.setflags(write=False)
+
+    @classmethod
+    def from_one_hot(
```
Comment (marvinpfoertner) [inline src/linpde_gp/linfuncops/diffops/_coefficients.py:27]:
```suggestion
    def from_index(
```

File: src/linpde_gp/linfuncops/diffops/_coefficients.py [hunk 9]
```diff
@@ -1,14 +1,61 @@
...
+        multi_index = np.zeros(shape, dtype=int)
+        multi_index[index] = order
+        return cls(multi_index)
+
+    @functools.cached_property
+    def total_order(self) -> int:
+        return np.sum(self._multi_index)
+
+    @property
+    def data(self) -> np.ndarray:
```
Comment (marvinpfoertner) [inline src/linpde_gp/linfuncops/diffops/_coefficients.py:39]:
```suggestion
    def array(self) -> np.ndarray:
```

**Review by marvinpfoertner — APPROVED** [review]
Beautiful! 🙏🏻

**Pull request merged**